[ONNX FE] Add SoftmaxCrossEntropyLoss operator to ONNX Frontend

AJThePro99 opened the pull request:
This pull request implements the SoftmaxCrossEntropyLoss operation for the ONNX frontend in OpenVINO, supporting opset 12 and opset 13. It includes new header and source files to define and implement the operation.

Ticket [#20547 ](https://github.com/openvinotoolkit/openvino/issues/20547)

New operation implementation:

* [`src/frontends/onnx/frontend/src/op/softmax_cross_entropy_loss.hpp`](diffhunk://#diff-9bdf7b47b5d5e0217813de8cb69984611733437e1f32b29820d175879457c49bR1-R22): Added the header file defining the `SoftmaxCrossEntropyLoss` operation for opset 12 and opset 13.
* [`src/frontends/onnx/frontend/src/op/softmax_crossentropy_loss.cpp`](diffhunk://#diff-6e4766d87154dfe88dcaa24e8e8568d6c14d63941e623b68e9f1b7f642053672R1-R77): Added the source file implementing the `SoftmaxCrossEntropyLoss` operation, including the shared helper function `impl_softmax_cross_entropy` to compute the loss.

---

#### TODO
- [x] 1. Create `softmax_cross_entropy_loss.cpp` & `softmax_cross_entropy_loss.hpp`
- [x] 2. Add implementation of [SoftMaxCrossEntropyLoss](https://github.com/onnx/onnx/blob/main/docs/Operators.md#softmaxcrossentropyloss)   for `opset_12` & `opset_13`
- [x] 3. Register the function in [ops_bridge.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/ops_bridge.cpp#L554-L560)
- [x] 4. Creating test models within [ONNX Models directory](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/tests/models/)
- [x] 5. Added test cases covering all use cases [here](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/onnx_import.in.cpp)
- [ ] 6. Check [Python xfailed tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/tests_python/test_backend.py) for added functionality
- [ ] 7. Final code formatting to adhere to OpenVINO codebase standards

---

P.S.: I'm very new to contributing to large codebases, and I want to know if my work meets the standards and most of all, is what's expected!
> As a new contributor, I would really appreciate your guidance on the following:

1. Does the implementation meet OpenVINO standards?
2. Are there additional edge cases or test scenarios I should include? (Excluding TODOs)
3. How can I improve code readability, efficiency, or maintainability?

AJThePro99 commented:
Hi @gkrivor 
I’ve created a draft pull request for the implementation of SoftmaxCrossEntropyLoss. Since I’m new to contributing to OpenVINO, I’d appreciate it if you could take a look and let me know if I’m on the right track. Let me know if there are any changes or additional requirements I should address. Thanks for your time!

AJThePro99 commented on src/frontends/onnx/frontend/src/op/softmax_crossentropy_loss.cpp:
```diff
@@ -0,0 +1,77 @@
+namespace frontend {
+namespace onnx {
+namespace {
+    // softmax cross entropy implementation (Shared helper fn)
```
outdated comment, please ignore. (I was trying the implementation with a `shared_ptr<>`

rkazants commented on src/frontends/onnx/frontend/src/op/softmax_crossentropy_loss.cpp:
```diff
@@ -0,0 +1,77 @@
+
+        const auto scores = inputs[0];
+        const auto labels = inputs[1];
+
```
I think `weights` are also needed according to the spec. Please check.

rkazants commented on src/frontends/onnx/frontend/src/op/softmax_crossentropy_loss.cpp:
```diff
@@ -0,0 +1,77 @@
+#include "openvino/op/negative.hpp"
+#include "openvino/op/reduce_mean.hpp"
+#include "openvino/op/reduce_sum.hpp"
+#include "openvino/op/select.hpp"
```
not all of these headers are needed, for example, select is not used. Please clean-up includes

AJThePro99 replied:
On it

AJThePro99 commented:
Hello!
I'd like to know if there exists a way to build the components of only the onnx frontend instead of the entire project `src/frontends/onnx`

It does have a CMakeLists.txt file, so I'm wondering if I could save time by doing that.
Or, does it not work that way?

AJThePro99 commented:
Please bear with me for a while. My laptop's number keys just gave up and do not function anymore. And I have no external keyboard at the moment. It's going to take a while for my next update, after the servicing is over.
I apologize for the delay

AJThePro99 commented:
Hello @rkazants !
I've successfully completed the implementation of the `SoftmaxCrossentropyLoss` function within `frontend/src/op`
And, I've started working on the test cases. 
My main question is how do you 'test' the functionality yourself?

AJThePro99 commented:
This pull request introduces the implementation of the Softmax Cross Entropy Loss operation for the ONNX frontend. The changes include adding new header and source files to define and implement the operation for two different opsets (12 and 13). The most important changes are as follows:

### Addition of new header and source files

* [`src/frontends/onnx/frontend/src/op/softmax_cross_entropy_loss.hpp`](diffhunk://#diff-9bdf7b47b5d5e0217813de8cb69984611733437e1f32b29820d175879457c49bR1-R22): Added a new header file to declare the `softmax_cross_entropy_loss` function for opset 12 and opset 13.
* [`src/frontends/onnx/frontend/src/op/softmax_crossentropy_loss.cpp`](diffhunk://#diff-6e4766d87154dfe88dcaa24e8e8568d6c14d63941e623b68e9f1b7f642053672R1-R117): Added a new source file to implement the `softmax_cross_entropy_loss` function, including the shared helper function `impl_softmax_cross_entropy` for the actual computation.

### Implementation details

* The `impl_softmax_cross_entropy` function handles the core logic of the Softmax Cross Entropy Loss, including support for optional weights and the ignore index attribute.
* The function supports reduction modes ("mean" and "sum") and includes error handling for dynamic rank scenarios.
---

I have a question about registering the operator.
According to [how_to_add_op.md](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/how_to_add_op.md#how-to-implement-a-new-operation-in-onnx-fe-codebase),  the registration looks like this within `ops_bridge.cpp`:
```cpp
#include "op/org.openvinotoolkit/custom_add.hpp"
...
REGISTER_OPERATOR_WITH_DOMAIN(OPENVINO_ONNX_DOMAIN, "CustomAdd", 1, custom_add);
```
But, [ops_bridge.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/ops_bridge.cpp) looks different.

But, when I compared `softmax_cross_entropy_loss.cpp` to other operators in the `onnx/frontend/src/op` directory, All of the operators are registered using a macro `ONNX_OP()`, hence - I used the same. This is just speculation for now, and I do not know if I properly completed the operator registration. 
I'd like for you to guide me in the right direction if there are more things to be done.

And, I'd really appreciate it if there is a resource that will help me write the `.prototxt` test model files. If you know any, please redirect me to them. It'll help me out a lot.

Thanks a bunch!

AJThePro99 commented:
Hi @gkrivor, Hi @rkazants , I’ve been working on this PR for a while now. I’d appreciate any feedback when you have time. Let me know if anything is blocking the review.

AJThePro99 commented:
Closing PR #28682 : Issues with ONNX tests, reopening a fresh PR.

gkrivor commented:
Hi @AJThePro99, could you re-open the PR?

gkrivor commented:
@AJThePro99 perfect! Thanks, I believe we will finally find a resources to continue for work on this PR. Sorry for a delay.
